### PR TITLE
support force load balance

### PR DESCRIPTION
**Pull request by yujuancao07**
support force load balance

## Purpose

In MoE, the router gate assigns tokens to experts unevenly. Each EP rank hosts a subset of routed experts, and the dispatch/combine all-to-all collectives move a volume of data proportional to the number of tokens that rank's local experts actually serve. Because real routing is noisy, per-rank collective volumes differ and a few ranks become stragglers. This routing noise also obscures the true cost of the MoE communication path, which makes compute/communication overlap hard to tune.
We need a switch that forces routing to a perfectly balanced, deterministic distribution, so every EP rank ships and receives an identical volume in the dispatch/combine collectives.

## Issue

- Related issue(s): #40 
- Closing keyword, only if fully resolved: Closes #

## Scope

- In scope:
- Out of scope:

## Docs Impact

- Files updated:
- If none, reason:

---

<details>
<summary>Essential PR Checklist</summary>

- [ ] Purpose is clear and linked to public context when possible.
- [ ] Scope is bounded.
- [ ] Compatibility with vLLM v0.19.1 is considered.
- [ ] No changes are made to the vLLM source checkout.
- [ ] Plugin-owned classes or explicit dotted class paths are preferred over monkey patches.
- [ ] Any compat shim or monkey patch is isolated, idempotent, version-guarded, documented, and tested.
- [ ] Imports remain CPU-safe; CUDA-heavy work is delayed or GPU-gated.
- [ ] Validation evidence is included, including skipped GPU tests when applicable.
- [ ] Documentation impact is stated.

</details>

### CODE REVIEW

File: afd_plugin/compat/patches/force_load_balance.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+                    rank * local_routed_experts,
+                    rank * local_routed_experts + self.force_load_balance_topn_per_rank,
+                    device=device,
+                    dtype=torch.int32,
+                )
+                for rank in range(self.ep_size)
+            ]
+            return torch.cat(per_rank_cycles, dim=0)
+
+        base = torch.randperm(self.n_routed_experts, device=device, dtype=torch.int32)
```
Comment (hsliuustc0106) [inline afd_plugin/compat/patches/force_load_balance.py]: Each EP rank will run this randperm independently, so ranks can build different fake routing tables for the same tokens. Use a deterministic shared order/seed so every rank agrees on topk_ids.

File: afd_plugin/compat/patches/force_load_balance.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        import torch
+
+        if self.force_lb_fake_topk_buffer is None:
+            raise RuntimeError("force_lb_fake_topk_buffer is not initialized")
+        if self.force_lb_fake_topk_buffer.device != device:
+            self.force_lb_fake_topk_buffer = self.force_lb_fake_topk_buffer.to(
+                device, non_blocking=True
+            )
+        if batch_tokens > self.force_lb_fake_topk_buffer.size(0):
+            return None
```
Comment (hsliuustc0106) [inline afd_plugin/compat/patches/force_load_balance.py]: Returning None silently disables force load balance for batches larger than the prebuilt buffer. Please grow/rebuild the buffer or fail clearly; otherwise the enabled mode can fall back to real, imbalanced routing.

File: afd_plugin/compat/patches/force_load_balance.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    if _W8A8_APPLY_BASELINE_MARKER not in apply_source:
+        logger.debug(
+            "AFD force-load-balance patch skipped: W8A8 apply baseline changed"
+        )
+        return
+
+    # ------------------------------------------------------------------ #
+    # Force-load-balance helpers (faithful to the upstream implementation). #
+    # ------------------------------------------------------------------ #
+    def _get_force_lb_max_tokens(vllm_config: Any) -> int:
```
Comment (hsliuustc0106) [inline afd_plugin/compat/patches/force_load_balance.py]: This patch puts most real Force-LB logic in nested private helpers typed as `Any`/`*args`/`**kwargs`. Please lift reusable pieces to module-level typed helpers or a small helper class, with documented inputs, so the patch remains reviewable and testable.

**Review by hsliuustc0106 — COMMENTED** [review]
Maintainability follow-up from the project review rules.

**Comment by hsliuustc0106** [comment]
1.每个文件最上方注释要规定文件的功能覆盖范围，核心的功能类&关键函数
2.每个类&函数的注释要定义清楚，入参要详细，杜绝kwargs, type any这些费解的入参和type；
3.utils类要能复用，杜绝功能文件里大面积私有util函数
4.关键类的函数要有详细设计，杜绝随意添加新函数，导致代码过于冗长，超出类函数范围

**Comment by yujuancao07** [comment]
image

**Comment by jiangkuaixue123** [comment]
这个patch文件应该 放到patches下面一个npu的文件夹里面 他只是npu用的 不能在GPU的环境中出现
